### PR TITLE
Move watcher history to data stream 

### PR DESCRIPTION
**Pull request by probakowski**
This change moves watcher history to data stream.

### CODE REVIEW

**Comment by probakowski** [comment]
@elasticmachine update branch

**Comment by probakowski** [comment]
I mark this as breaking since we change naming convention of watcher history - it no longer contains date in it, there's single data stream (per version) for all records

**Comment by sebelga** [comment]
@probakowski do you mind elaborating with an example of the difference in the response (before, after) so we can better assess if this impacts our UI? cheers!

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/history/HistoryStore.java [hunk 1]
```diff
@@ -78,9 +76,12 @@ public void forcePut(WatchRecord watchRecord) {
...
-        String currentIndex = HistoryStoreField.getHistoryIndexNameForTime(ZonedDateTime.now(ZoneOffset.UTC));
-        IndexMetadata indexMetadata = WatchStoreUtils.getConcreteIndex(currentIndex, state.metadata());
+        IndexMetadata indexMetadata = WatchStoreUtils.getConcreteIndex(HistoryStoreField.DATA_STREAM, state.metadata());
         return indexMetadata == null || (indexMetadata.getState() == IndexMetadata.State.OPEN &&
             state.routingTable().index(indexMetadata.getIndex()).allPrimaryShardsActive());
     }
+
+    public void flush() {
+        bulkProcessor.flush();
+    }
```
Comment (danhermann) [inline x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/history/HistoryStore.java:86]: Out of curiosity, is this addressing a scenario that's specific to data streams or just a general case where history entries weren't being written?

**Review by danhermann — APPROVED** [review]
Thanks, @probakowski. Looks pretty good to me. I left one minor comment to address and one minor question.

**Comment by probakowski** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-windows

**Comment by probakowski** [comment]
@elasticmachine run elasticsearch-ci/2

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/Watcher.java [hunk 2]
```diff
@@ -615,14 +615,7 @@ static void validAutoCreateIndex(Settings settings, Logger logger) {
...
         ZonedDateTime now = ZonedDateTime.now(ZoneOffset.UTC);
-        indices.add(HistoryStoreField.getHistoryIndexNameForTime(now));
-        indices.add(HistoryStoreField.getHistoryIndexNameForTime(now.plusDays(1)));
-        indices.add(HistoryStoreField.getHistoryIndexNameForTime(now.plusMonths(1)));
-        indices.add(HistoryStoreField.getHistoryIndexNameForTime(now.plusMonths(2)));
-        indices.add(HistoryStoreField.getHistoryIndexNameForTime(now.plusMonths(3)));
-        indices.add(HistoryStoreField.getHistoryIndexNameForTime(now.plusMonths(4)));
-        indices.add(HistoryStoreField.getHistoryIndexNameForTime(now.plusMonths(5)));
-        indices.add(HistoryStoreField.getHistoryIndexNameForTime(now.plusMonths(6)));
+        indices.add(HistoryStoreField.DATA_STREAM);
```
[the remark this answers is not in the document]
Comment (probakowski) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/Watcher.java]: Thanks, I've removed it

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/history/HistoryStore.java (hunk 1 above)
Comment (probakowski) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/history/HistoryStore.java:86]: Main reason for this is to have clear state after tests. This makes sure that we don't have any pending docs to index so we can do proper cleanup after the test and leave no hanging indices behind

**Comment by probakowski** [comment]
@sebelga we've changed naming convention here, previously index name contained date it was created (so record created today would go to index `.watcher-history-13-2020-11-12`) but now everything goes to data stream called `.watcher-history-14` (template version bump). So, as long as you use `.watcher-history-*` for search there's no difference but if your workflow queries specific index (like 3rd party software or watcher with input pointing to name with date math instead of a wildcard) it will break

**Comment by sebelga** [comment]
@probakowski was this message for me or @sebgl ?

**Comment by cjcenizal** [comment]
Tested Watcher UI with these changes and it seems to work as expected. I looked at the code and we're retrieving Watcher history with a search API request aimed at `.watcher-history-*, using `scroll`, `sort`, `size`, and `query` parameters, all of which I believe are still supported. Thanks for the notes, @probakowski!

**Pull request merged**